selfdrive/updated.py

```diff
@@ -219,6 +223,58 @@ def finalize_update() -> None:
   cloudlog.info("done finalizing overlay")
 
 
+def download(filename, url, img_hash):
+  decompressor = lzma.LZMADecompressor(format=lzma.FORMAT_AUTO)
+  sha256 = hashlib.sha256()
+  with requests.get(url, stream=True, headers={'Accept-Encoding': None}) as stream:
+    stream.raise_for_status()
+    with open(filename, 'wb') as f:
+      for chunk in stream.iter_content(chunk_size=1024*1024):
+        decompressed_chunk = decompressor.decompress(chunk)
+        sha256.update(decompressed_chunk)
+        f.write(decompressed_chunk)
+
+  if sha256.hexdigest().lower() != img_hash.lower():
+    raise Exception(f"'{filename}' hash mismatch, got {sha256.hexdigest().lower()}")
+
+
+def handle_agnos_update(wait_helper):
+  with open(NEOS_VERSION, "r") as f:
+    cur_version = f.read().strip()
+
+  updated_version = run(["bash", "-c", r"unset AGNOS_VERSION && source launch_env.sh && \
+                       echo -n $AGNOS_VERSION"], OVERLAY_MERGED).strip()
+
+  cloudlog.info(f"AGNOS version check: {cur_version} vs {updated_version}")
+  if cur_version == updated_version:
+    return
+
+  cloudlog.info(f"Beginning background installation for AGNOS {updated_version}")
+
+  update = json.load(open(os.path.join(OVERLAY_MERGED, "installer/updater/update_agnos.json")))
+
+  current_slot = subprocess.check_output(["abctl", "--boot_slot"], encoding='utf-8').strip()
+  target_slot = "_b" if current_slot == "_a" else "_a"
+  target_slot_number = "0" if target_slot == "_a" else "1"
+
+  # set target slot as unbootable
+  #os.system(f"abctl --set_unbootable {target_slot_number}")
+
+  for partition in update:
+    img_path = f"/tmp/{partition['name']}_{partition['hash']}.img"
```

Comment:
Isn't `/tmp` too small to hold the system partition image?
Also, the uncompressed image is still sparse, which doesn't seem to be handled. Does this work?
We might need to stream it directly into the partition, not save the image first